### ISSUE TITLE
ElasticSearch dynamic templates not work

### ISSUE THREAD

**Issue by CJSoldier** [JanusGraph/janusgraph#725]
It's a new problem after #699.

 The mappings "being overrided" after I build mixed index, though I‘ve’ already set `elasticsearch.create.use-external-mappings` as true.

When we use external mappings with dynamic templates configured, we expect ES generate mappings automatically, right?

```
if (useExternalMappings) {
            try {
                //We check if the externalMapping have the property 'key'
                final IndexMapping mappings = client.getMapping(indexStoreName, store);
                if (mappings == null || (!mappings.isDynamic() && !mappings.getProperties().containsKey(key))) {
                    //Error if it is not dynamic and have not the property 'key'
                    throw new PermanentBackendException("The external mapping for index '"+ indexStoreName + "' and type '" + store + "' do not have property '" + key + "'");
                } else if (mappings.isDynamic()) {
                    //If it is dynamic, we push the unknown property 'key'
                    // this.pushMapping(store, key, information);
                }
            } catch (final IOException e) {
                throw new PermanentBackendException(e);
            }
        }
```
So I commented this line `this.pushMapping(store, key, information);` in org.janusgraph.diskstorage.es.ElasticSearchIndex.java.

After doing that, the result is what I want. I just want to customize _all field with some fields excluded.
Should we comment that line?

**Comment by davidclement90** [comment]
I code like this if the user would like to use external mapping not for all properties.
But to solve your and my case, I think we should not comment this line but change the condition (isDynamic and is not using dynamic template) instead of just isDynamic
@sjudeng what do you think ?

**Comment by sjudeng** [comment]
@davidclement90 Do you have a use case for managing the mapping externally for only select properties? If not I'd just use the single property (`use-external-mappings`) so then the user just has the option to either manage the mapping externally or let JanusGraph manage it, but not both.

**Comment by davidclement90** [comment]
When I just want to change the mapping of one or two properties among many properties. And when I add another property to the index latter, I do not have to set the mapping in ES.
 It's easier to do like this but I can set the entire mapping.

**Comment by CJSoldier** [comment]
@davidclement90, I understand that.

The thing is, what if a new mixed index also matches external mappings(dynamic template)?
If then, I'm afraid it's hard to code, because we don't know whether the new property will match external mappings.

Dynamic tempate in ES is just like dynamic fields in Solr. When using that, we could cover all the possible scenarios.

**Comment by davidclement90** [comment]
Your are right.
Maybe just add another key like use-dynamic-template default false
If false then push the mapping (for my case) else do not push (for your case) ?

**Comment by CJSoldier** [comment]
Yeah, I agree. I will configure dynamic template carefully and set use-dynamic-template as true.
